src/natives/java_util.ts

```diff
@@ -307,35 +307,126 @@ class java_util_zip_CRC32 {
   }
 
-  public static 'deflateBytes(J[BIII)I'(thread: JVMThread, javaThis: JVMTypes.java_util_zip_Deflater, arg0: Long, arg1: JVMTypes.JVMArray<number>, arg2: number, arg3: number, arg4: number): number {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
+  public static 'deflateBytes(J[BIII)I'(thread: JVMThread, javaThis: JVMTypes.java_util_zip_Deflater, addr: Long, b: JVMTypes.JVMArray<number>, off: number, len: number, flush: number): number {
+    let strm = GetZStream(thread, addr.toNumber());
+    if (!strm) return;
+
+    let thisBuf = javaThis['java/util/zip/Deflater/buf'];
+    let thisOff = javaThis['java/util/zip/Deflater/off'];
+    let thisLen = javaThis['java/util/zip/Deflater/len'];
+
+    let inBuf = thisBuf.array;
+    let outBuf = b.array;
+
+    strm.input = i82u8(inBuf, 0, inBuf.length);
+    strm.next_in = thisOff;
+    strm.avail_in = thisLen;
+
+    strm.output = i82u8(outBuf, 0, outBuf.length);
+    strm.next_out = off;
+    strm.avail_out = len;
+
+    if (javaThis['java/util/zip/Deflater/setParams']) {
+      let level = javaThis['java/util/zip/Deflater/level'];
+      let strategy = javaThis['java/util/zip/Deflater/level'];
+      //deflateParams is not yet supported by pako. We'll open a new ZStream with the new parameters instead.
+      // res = deflate.deflateParams(strm, level, strategy); 
+      let newStream = new ZStreamCons();
+      let res = deflate.deflateInit2(newStream, strm.state.level, strm.state.method, strm.state.windowBits, strm.state.memLevel, strm.state.strategy);
+      ZStreams[addr.toNumber()] = newStream;
+      switch (res) {
+        case ZlibReturnCode.Z_OK:
+          javaThis['java/util/zip/Deflater/setParams'] = 0;
+          thisOff += thisLen - strm.avail_in;
+          javaThis['java/util/zip/Deflater/off'] = thisOff;
+          javaThis['java/util/zip/Deflater/len'] = strm.avail_in;
+          return len - strm.avail_out;
+        case ZlibReturnCode.Z_BUF_ERROR:
+          javaThis['java/util/zip/Deflater/setParams'] = 0;
+          return 0;
+        default:
+          thread.throwNewException("Ljava/lang/InternalError;", strm.msg);
+      }
+    } else {
+      let finish = javaThis['java/util/zip/Deflater/finish'];
+
+      let res = deflate.deflate(strm, finish ? ZlibFlushValue.Z_FINISH : flush);
+
+      switch (res) {
+        case ZlibReturnCode.Z_STREAM_END:
+          javaThis['java/util/zip/Deflater/finished'] = 1;
+          // intentionally fall through
+        case ZlibReturnCode.Z_OK:
+          thisOff += thisLen - strm.avail_in;
+          javaThis['java/util/zip/Deflater/off'] = thisOff;
+          javaThis['java/util/zip/Deflater/len'] = strm.avail_in;
+          return len - strm.avail_out;
+        case ZlibReturnCode.Z_BUF_ERROR:
+          return 0;
+        default:
+          thread.throwNewException('Ljava/lang/InternalError;', strm.msg);
+      }
+    }
     return 0;
```

Comment:
You should remove this. If this function throws an exception, it'll return twice!